### ISSUE TITLE
Why use  (tanh(a + b)) instead of  (tanh(a) + tanh(b)) or does it not matter in some cases?

### ISSUE THREAD

**Issue by UtpalMattoo** [tensorflow/tensorflow#45379]
I am not sure if this is a documentation issue or an actual problem - please advise:

In this [machine translation exercise](https://www.tensorflow.org/tutorials/text/nmt_with_attention):

Why does this `class BahdanauAttention` (see class definition at end) do the additive attention as:

`tf.nn.tanh(self.W1(query_with_time_axis) + self.W2(values))`

instead of:

`tf.nn.tanh(self.W1(query_with_time_axis) + tf.nn.tanh(self.W1(query_with_time_axis)`

Or, in other words, why does it use: `tanh(a+b)` instead of `tanh(a) + tanh(b)`?

If both `self.W1` and `self.W2` are separate neural network layers, why don't they get their own separate activation functions?

`tanh` restricts the output between `-1` and `+1`, centered at 0 which makes learning gradient descent quicker

```
class BahdanauAttention(tf.keras.layers.Layer):
  def __init__(self, units):
    super(BahdanauAttention, self).__init__()
    self.W1 = tf.keras.layers.Dense(units)
    self.W2 = tf.keras.layers.Dense(units)
    self.V = tf.keras.layers.Dense(1)

  def call(self, query, values):
    # query hidden state shape == (batch_size, hidden size)
    # query_with_time_axis shape == (batch_size, 1, hidden size)
    # values shape == (batch_size, max_len, hidden size)
    # we are doing this to broadcast addition along the time axis to calculate the score
    query_with_time_axis = tf.expand_dims(query, 1)

    # score shape == (batch_size, max_length, 1)
    # we get 1 at the last axis because we are applying score to self.V
    # the shape of the tensor before applying self.V is (batch_size, max_length, units)
    score = self.V(tf.nn.tanh(
        self.W1(query_with_time_axis) + self.W2(values)))

    # attention_weights shape == (batch_size, max_length, 1)
    attention_weights = tf.nn.softmax(score, axis=1)

    # context_vector shape after sum == (batch_size, hidden_size)
    context_vector = attention_weights * values
    context_vector = tf.reduce_sum(context_vector, axis=1)

    return context_vector, attention_weights


```

**Comment by foxik** [comment]
Questions unrelated to TensorFlow are better suited for general Q&A sites like https://stackoverflow.com/ .

Personally, I think that both approaches would result in similar performance; the idea behind the original formulation is that you project both the decoder state and the encoded word to the same space (using a linear transformation), sum them, and then continue with a single representation (so apply bias, non-linearity, and process the result further). But you can also imagine the projection to the shared space to be non-linear (including the tanh).

Mathematically, the biggest difference between the two formulations is when for example a=-4 and b=2; then the individual tanhs are close to -1 and 1, so the sum is 0; but tanh of the sum is close to -1 -- but if the outputs of the previous layers would be small, then tanh(a) is close to a, and tanh(a+b) ~ a + b ~ tanh(a) + tanh(b).

**Comment by UtpalMattoo** [comment]
Yes, thanks for clarifying this. I was curious if this was a possible code change so posted it here as well.

Makes sense that the difference of the outcome between the two approaches (`tanh(a+b)` VS `tanh(a) + tanh(b)`) will be high when `a` (or `query_with_time_axis`) and `b` (`values` or the encoder output) are each in the plateau portion of the `tanh`. Otherwise, generally similar performance.
